File: backend/app/models/source.py
# app/models/source.py
# Purpose: Pydantic schemas for Source management.
# Responsibilities: Defines validation models for Source metadata and requests.

from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional

class SourceBase(BaseModel):
    name: str = Field(..., description="Name of the source file, video title, or website title")
    type: str = Field(..., description="Type of source (pdf, image, audio, youtube, website, text, email)")

class YouTubeCreate(BaseModel):
    url: str = Field(..., description="YouTube video URL")

class WebsiteCreate(BaseModel):
    url: str = Field(..., description="Website page URL")

class TextCreate(BaseModel):
    name: str = Field(..., description="Name for the copied text source")
    content: str = Field(..., description="Raw text content to ingest")

class EmailCreate(BaseModel):
    subject: str = Field(..., description="Subject of the email")
    sender: str = Field(..., description="Sender/From of the email")
    recipient: str = Field(..., description="Recipient/To of the email")
    body: str = Field(..., description="Body text content of the email")

class Source(SourceBase):
    id: str = Field(..., description="Unique source identifier")
    path: Optional[str] = Field(None, description="Relative path of saved file on disk")
    url: Optional[str] = Field(None, description="YouTube link if type is youtube")
    added_at: datetime = Field(..., description="Timestamp when source was attached")
    size_bytes: Optional[int] = Field(None, description="Size of file in bytes")
    status: str = Field("pending", description="Processing status (pending, processing, ready, failed)")
    summary: Optional[str] = Field(None, description="Extracted textual summary or preview")
    stats: Optional[dict] = Field(None, description="Extracted statistics (pages, words, chunks)")

    class Config:
        json_schema_extra = {
            "example": {
                "id": "s7f8d223-9f8e-4a6c-905c-e6593539a2f1",
                "name": "annual_report.pdf",
                "type": "pdf",
                "path": "storage/workspaces/e4f8d223/sources/annual_report.pdf",
                "url": None,
                "added_at": "2026-06-16T22:15:00.000Z",
                "size_bytes": 1048576,
                "status": "pending"
            }
        }

File: frontend/lib/features/source_upload/models/source.dart
// features/source_upload/models/source.dart
// Purpose: Defines the Source model and source type enum.
// Responsibilities: Handles JSON serialization and type mapping.

enum SourceType {
  pdf,
  image,
  audio,
  youtube,
  website,
  text,
  email;

  static SourceType fromString(String val) {
    switch (val.toLowerCase()) {
      case 'image':
        return SourceType.image;
      case 'audio':
        return SourceType.audio;
      case 'youtube':
        return SourceType.youtube;
      case 'website':
        return SourceType.website;
      case 'text':
        return SourceType.text;
      case 'email':
        return SourceType.email;
      case 'pdf':
      default:
        return SourceType.pdf;
    }
  }

  String toJson() => name;
}

enum SourceStatus {
  pending,
  processing,
  ready,
  failed;

  static SourceStatus fromString(String val) {
    switch (val.toLowerCase()) {
      case 'processing':
        return SourceStatus.processing;
      case 'ready':
        return SourceStatus.ready;
      case 'failed':
        return SourceStatus.failed;
      case 'pending':
      default:
        return SourceStatus.pending;
    }
  }

  String toJson() => name;
}

class Source {
  final String id;
  final String name;
  final SourceType type;
  final String? path;
  final String? url;
  final DateTime addedAt;
  final int? sizeBytes;
  final SourceStatus status;
  final String? summary;
  final Map<String, dynamic>? stats;

  Source({
    required this.id,
    required this.name,
    required this.type,
    this.path,
    this.url,
    required this.addedAt,
    this.sizeBytes,
    required this.status,
    this.summary,
    this.stats,
  });

  factory Source.fromJson(Map<String, dynamic> json) {
    return Source(
      id: json['id'] as String,
      name: json['name'] as String,
      type: SourceType.fromString(json['type'] as String),
      path: json['path'] as String?,
      url: json['url'] as String?,
      addedAt: DateTime.parse(json['added_at'] as String),
      sizeBytes: json['size_bytes'] as int?,
      status: SourceStatus.fromString(json['status'] as String? ?? 'pending'),
      summary: json['summary'] as String?,
      stats: json['stats'] != null ? Map<String, dynamic>.from(json['stats'] as Map) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'type': type.toJson(),
      'path': path,
      'url': url,
      'added_at': addedAt.toIso8601String(),
      'size_bytes': sizeBytes,
      'status': status.toJson(),
      'summary': summary,
      'stats': stats,
    };
  }
}
